cli/src/main.rs

```diff
@@ -0,0 +1,303 @@
+                    let funnel_url = ensure_funnel_running().await;
+                    config.set_base_path(&funnel_url);
+                }
+
+                match TES::new(&config).await {
+                    Ok(tes) => {
+                        let task = tes.get(view, id).await;
+                        println!("{:?}",task);
+                    },
+                    Err(e) => {
+                        println!("Error creating TES instance: {:?}", e);
+                        return Err(e);
+                    }
+                };
+            }
+            if let Some(("status", sub)) = sub.subcommand() {   
+                let id = sub.value_of("id").unwrap().to_string();
+                
+                let mut config = load_configuration();   
+                if config.base_path == "localhost" {
+                    let funnel_url = ensure_funnel_running().await;
+                    config.set_base_path(&funnel_url);
+                }
+                let transport = Transport::new(&config);
+                let task = Task::new(id, transport);
+                match task.status().await {
+                    Ok(status) => {
+                        println!("The status is: {:?}",status);
+                    },
+                    Err(e) => {
+                        println!("Error creating Task instance: {:?}", e);
+                        return Err(e);
+                    }
+                };
+            }
+            if let Some(("cancel", sub)) = sub.subcommand() {   
+                let id = sub.value_of("id").unwrap().to_string();
+                
+                let mut config = load_configuration();
+                if config.base_path == "localhost" {
+                    let funnel_url = ensure_funnel_running().await;
+                    config.set_base_path(&funnel_url);
+                }
+                let transport = Transport::new(&config);
+                let task = Task::new(id, transport);
+                match task.cancel().await {
+                    Ok(output) => {
+                        println!("The new value is: {:?}",output);
+                    },
+                    Err(e) => {
+                        println!("Error creating Task instance: {:?}", e);
+                        return Err(e);
+                    }
+                };
+            }
+        }
+        
+        _ => {
+            eprintln!("Error: Unrecognized command or option");
+            std::process::exit(1);
+        }
+    }
+    Ok(())
+}
+
+/// Example `config.json` file:
+///
+/// ```json
+/// {
+///   "base_path": "http://localhost:8000",
+///   "user_agent": "username"
+/// }
+/// ```
+
+fn read_configuration_from_file(file_path: &str) -> Result<Configuration, Box<dyn Error>> {
+    let mut file = File::open(file_path)?;
+    let mut contents = String::new();
+    file.read_to_string(&mut contents)?;
+    
+    let json_value: Value = serde_json::from_str(&contents)?;
```

Comment:
as the SDK/CLI works with JSON data in APIs it handles, it might also be the optimal way to store config as JSON. Not sure if it's best for user experience. For the reference: AWS CLI – INI format,  Google Cloud CLI – text-based files with key-value pairs, Azure CLI – JSON format, IBM Cloud CLI – YAML format.

Reply:
I initially had tried implementing it like that only (of course, it is the neatest way to solve the problem), but the problem is that I think some data types in Configuration Struct are not deserializable, for eg the Url crate. If you know of a workaround, or are able to do this, please do tell me, and I will look into it again.

Reply:
For the specific problem, maybe this? https://docs.rs/url_serde/latest/url_serde/

As for how to store the config: JSON sounds good to me. Or why not use YAML? It's easier to manipulate, is a superset of JSON and can almost always be directly converted to JSON.

Reply:
Changed